src/main/java/com/google/codeu/servlets/CountryServlet.java

```diff
@@ -71,7 +90,13 @@ public void doGet(HttpServletRequest request, HttpServletResponse response)
     request.setAttribute("code", countryCode);
     request.setAttribute("name", countryData.getName());
     request.setAttribute("messages", messages);
+    request.setAttribute("countryCode", countryCode);
+    request.setAttribute("category", category);
     request.setAttribute("isUserLoggedIn", userService.isUserLoggedIn());
-    request.getRequestDispatcher("/WEB-INF/country.jsp").forward(request, response);
+    if(category != null) {
+      request.getRequestDispatcher("/c").forward(request, response);
+    } else {
+      request.getRequestDispatcher("/WEB-INF/country.jsp").forward(request, response);
```

Comment:
Should we log an error here? What does it mean if we don't have a valid country format? How does the page load?

Reply:
I'm not sure I understand, by country format, do you mean what's in the URL? What exactly would be the error?

Reply:
So I guess what I'm asking is like "what does it mean if category is null?" are we just on a country page, or did someone enter a category wrong?

Reply:
I think I read the code backwards; I think this is good, So it goes to the category page if the category exists, and loads the country page if it does not?

Reply:
Yeah! I set category to null at the top, and it only gets set if there is a "/c" in the url, between the country and the category, in which it falls into that block of code. I'll ask Aaron if there's a better way to handle the url, because I think this is kind of a hacky way to do it haha, but yeah, if it's not set, then that means the url didn't include /c , and that it is just accessing the country home page. If the category didn't exist, then CategoryServlet redirects to "/invalid-category/"

Reply:
Ooh, once again, fights with code that took too long might come to the rescue! Do you know regex? You can use it to see if it's present, and then search the url for the thing. I did that in Python though so I'm not completely sure of Java syntax, but I think @ArgaAlan used it in one of his [commits](https://github.com/ArgaAlan/CodeU-2019/commit/9ce2d92670971e1d64d0108263f9c8aa87e68780). 
So [here's ](https://www.tutorialspoint.com/javaregex/javaregex_capturing_groups.htm) some boring information on them, and there are regex testers like regex101 but I can't find one for Java. There's a bunch of weird syntax things but basically it's fancy search-and-find, and so I think you can do something like 
```
String regexUrl = "\/c\/(.*)"; 
```
This is "/c/" but you have to escape the "/". It will "capture" everything that comes after,  
```
Pattern urlPattern = Pattern.compile(regexUrl);
Matcher urlMatcher = urlPattern.matcher(url);
```
This makes a Matcher object with information about whether or not the pattern was found and if so, what the "captured" stuff is
```
if (urlMatcher.find()) {
     //category exists, so send to page
    response.sendRedirect("/"+country+"/c/"+urlMatcher(0));
    return;
}
//else load country page
```

**something like that? I know my syntax is probably off but hopefully that gets you close?**
*Regex was the bane of my existence on co-op, we used it to like find senders and stuff from the email. So if you have any questions, I have worked with it...far...too...much. (And I still don't know much, smh)*

Reply:
It's already the bane of my existence and I haven't even worked with it xD Thank you for walking me through it though, I think I get the gist of it!